File: hw2/task4/python/SaveTheTribe.py
"""
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Chao Yu
 * Student ID : 1155053722
 * Email Addr : ychao5@cse.cuhk.edu.hk
"""

import random

from Map import Map
from Soldier import Soldier
from Spring import Spring
from Monster import Monster

class SaveTheTribe():
    def __init__(self):
        self.map = Map()
        self.soldier = Soldier()
        self.spring = Spring()
        self.monsters = []
        self.game_enabled = True
    
    def initialize(self):
        self.monsters.append(Monster(1, random.randint(0, 4) * 10 + 30))
        self.monsters[0].set_pos(4, 1)
        self.monsters[0].add_drop_item(2)
        self.monsters[0].add_drop_item(3)

        self.monsters.append(Monster(2, random.randint(0, 4) * 10 + 30))
        self.monsters[1].set_pos(3, 3)
        self.monsters[1].add_drop_item(3)
        self.monsters[1].add_drop_item(6)
        self.monsters[1].add_hint(1)
        self.monsters[1].add_hint(5)

        self.monsters.append(Monster(3, random.randint(0, 4) * 10 + 30))
        self.monsters[2].set_pos(5, 3)
        self.monsters[2].add_drop_item(4)
        self.monsters[2].add_hint(1)
        self.monsters[2].add_hint(2)

        self.monsters.append(Monster(4, random.randint(0, 4) * 10 + 30))
        self.monsters[3].set_pos(5, 5)
        self.monsters[3].add_hint(3)
        self.monsters[3].add_hint(6)

        self.monsters.append(Monster(5, random.randint(0, 4) * 10 + 30))
        self.monsters[4].set_pos(1, 4)
        self.monsters[4].add_drop_item(2)
        self.monsters[4].add_drop_item(6)

        self.monsters.append(Monster(6, random.randint(0, 4) * 10 + 30))
        self.monsters[5].set_pos(3, 5)
        self.monsters[5].add_drop_item(4)
        self.monsters[5].add_drop_item(7)
        self.monsters[5].add_hint(2)
        self.monsters[5].add_hint(5)

        self.monsters.append(Monster(7, random.randint(0, 4) * 10 + 30))
        self.monsters[6].set_pos(4, 7)
        self.monsters[6].add_drop_item(-1)
        self.monsters[6].add_hint(6)

        self.map.add_object(self.monsters)

        self.soldier.set_pos(1, 1)
        self.soldier.add_key(1)
        self.soldier.add_key(5)

        self.map.add_object(self.soldier)

        self.spring.set_pos(7, 4)

        self.map.add_object(self.spring)
    
    def start(self):
        print("=> Welcome to the desert!")
        print("=> Now you have to defeat the monsters and find the artifact to save the tribe.\n")

        while self.game_enabled:
            self.map.display_map()
            self.soldier.display_information()

            move = input("\n=> What is the next step? (W = Up, S = Down, A = Left, D = Right.) Input: ")
            pos = self.soldier.get_pos()
            old_row, old_column = pos.get_row(), pos.get_column()
            new_row, new_column = pos.get_row(), pos.get_column()
            
            if move.lower() == "w":
                new_row = old_row - 1
            elif move.lower() == "s":
                new_row = old_row + 1
            elif move.lower() == "a":
                new_column = old_column - 1
            elif move.lower() == "d":
                new_column = old_column + 1
            else:
                print("=> Illegal move!\n")
                continue
        
            if self.map.check_move(new_row, new_column):
                occupied_object = self.map.get_occupied_object(new_row, new_column)

                if occupied_object != None:
                    occupied_object.action_on_soldier(self.soldier)
                else:
                    self.soldier.move(new_row, new_column)
                    self.map.update(self.soldier, old_row, old_column, new_row, new_column)
                    print("\n")
            else:
                print("=> Illegal move!\n")
            
            if self.soldier.get_health() <= 0:
                print("=> You died.")
                print("=> Game over.\n")
                self.game_enabled = False
            
            if -1 in self.soldier.get_keys():
                print("=> You found the artifact.")
                print("=> Game over.\n")
                self.game_enabled = False

if __name__  == "__main__":
    game = SaveTheTribe()
    game.initialize()
    game.start()

File: hw2/task4/java/SaveTheTribe.java
/*
* CSCI3180 Principles of Programming Languages
*
* --- Declaration ---
*
* I declare that the assignment here submitted is original except for source
* material explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
*
* Assignment 2
* Name : Chao Yu
* Student ID : 1155053722
* Email Addr : ychao5@cse.cuhk.edu.hk
*/

import java.util.Random;
import java.util.Scanner;

public class SaveTheTribe {
  private Map map;
  private Soldier soldier;
  private Spring spring;
  private Monster[] monsters;

  private boolean gameEnabled;

  public SaveTheTribe() {
    this.map = new Map();
    this.soldier = new Soldier();
    this.spring = new Spring();
    this.monsters = new Monster[7];
    this.gameEnabled = true;
  }

  public void initialize() {
    /*
     * We use the integer 1-7 to represent the keys for corresponding caves, and the
     * integer -1 to represent the artifact.
     */

    Random random = new Random();

    this.monsters[0] = new Monster(1, random.nextInt(5) * 10 + 30);
    this.monsters[0].setPos(4, 1);
    this.monsters[0].addDropItem(2);
    this.monsters[0].addDropItem(3);

    this.monsters[1] = new Monster(2, random.nextInt(5) * 10 + 30);
    this.monsters[1].setPos(3, 3);
    this.monsters[1].addDropItem(3);
    this.monsters[1].addDropItem(6);
    this.monsters[1].addHint(1);
    this.monsters[1].addHint(5);

    this.monsters[2] = new Monster(3, random.nextInt(5) * 10 + 30);
    this.monsters[2].setPos(5, 3);
    this.monsters[2].addDropItem(4);
    this.monsters[2].addHint(1);
    this.monsters[2].addHint(2);

    this.monsters[3] = new Monster(4, random.nextInt(5) * 10 + 30);
    this.monsters[3].setPos(5, 5);
    this.monsters[3].addHint(3);
    this.monsters[3].addHint(6);

    this.monsters[4] = new Monster(5, random.nextInt(5) * 10 + 30);
    this.monsters[4].setPos(1, 4);
    this.monsters[4].addDropItem(2);
    this.monsters[4].addDropItem(6);

    this.monsters[5] = new Monster(6, random.nextInt(5) * 10 + 30);
    this.monsters[5].setPos(3, 5);
    this.monsters[5].addDropItem(4);
    this.monsters[5].addDropItem(7);
    this.monsters[5].addHint(2);
    this.monsters[5].addHint(5);

    this.monsters[6] = new Monster(7, random.nextInt(5) * 10 + 30);
    this.monsters[6].setPos(4, 7);
    this.monsters[6].addDropItem(-1);
    this.monsters[6].addHint(6);

    this.map.addObject(monsters);

    this.soldier.setPos(1, 1);
    this.soldier.addKey(1);
    this.soldier.addKey(5);

    this.map.addObject(this.soldier);

    this.spring.setPos(7, 4);

    this.map.addObject(this.spring);
  }

  public void start() {
    System.out.printf("=> Welcome to the desert!%n");
    System.out.printf("=> Now you have to defeat the monsters and find the artifact to save the tribe.%n%n");

    Scanner sc = new Scanner(System.in);

    while (gameEnabled) {
      this.map.displayMap();
      this.soldier.displayInformation();

      System.out.printf("%n=> What is the next step? (W = Up, S = Down, A = Left, D = Right.) Input: ");

      String move = sc.nextLine();

      Pos pos = this.soldier.getPos();
      int newRow, oldRow;
      int newColumn, oldColumn;

      newRow = oldRow = pos.getRow();
      newColumn = oldColumn = pos.getColumn();

      if (move.equalsIgnoreCase("W")) {
        newRow = oldRow - 1;
      } else if (move.equalsIgnoreCase("S")) {
        newRow = oldRow + 1;
      } else if (move.equalsIgnoreCase("A")) {
        newColumn = oldColumn - 1;
      } else if (move.equalsIgnoreCase("D")) {
        newColumn = oldColumn + 1;
      } else {
        System.out.printf("=> Illegal move!%n%n");
        continue;
      }

      if (this.map.checkMove(newRow, newColumn)) {
        Object occupiedObject = this.map.getOccupiedObject(newRow, newColumn);

        if (occupiedObject instanceof Monster) {
          ((Monster) occupiedObject).actionOnSoldier(this.soldier);
        } else if (occupiedObject instanceof Spring) {
          ((Spring) occupiedObject).actionOnSoldier(this.soldier);
        } else {
          this.soldier.move(newRow, newColumn);
          this.map.update(this.soldier, oldRow, oldColumn, newRow, newColumn);
          System.out.printf("%n%n");
        }
      } else {
        System.out.printf("=> Illegal move!%n%n");
      }

      if (this.soldier.getHealth() <= 0) {
        System.out.printf("=> You died.%n");
        System.out.printf("=> Game over.%n%n");
        this.gameEnabled = false;
      }

      /* Check if the soldier has received the artifact. */
      if (this.soldier.getKeys().contains(-1)) {
        System.out.printf("=> You found the artifact.%n");
        System.out.printf("=> Game over.%n%n");
        this.gameEnabled = false;
      }
    }
  }

  public static void main(String[] args) {
    SaveTheTribe game = new SaveTheTribe();
    game.initialize();
    game.start();
  }
}